From 3f1aa0d44407655880a363e87fe62209bc7b16b7 Mon Sep 17 00:00:00 2001
From: Tamal <tasaha@habana.ai>
Date: Thu, 08 Aug 2024 19:22:45 +0300
Subject: [PATCH] [SW-161727] g3: update bist result post ecc init

This patch updates die 0 sram bist result after ecc
init to boot info in ppboot stage.

This patch replaces boot_info_set_no_merge api to
ppboot_boot_info_set api invoke for die-0 sram bist
result in ppboot stage.

Change-Id: I47182143b4528069a083f7fcb3af38b6eef94c51
Signed-off-by: Tamal <tasaha@habana.ai>
---

diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
index e5a9c62..0a37c41 100644
--- a/arch/gaudi3/arc/asic.c
+++ b/arch/gaudi3/arc/asic.c
@@ -56,15 +56,6 @@
 		/* TODO: update "skip" by "pass" after memcpy engine timeout resolved */
 		preboot_puts("Die 0 cache memories MBIST Pass\n");
 	}
-
-	/*
-	 * Store Ppboot MBIST result in boot info
-	 */
-	if (boot_info_set_no_merge(HL_BOOT_INFO_SRAM_BIST_RESULT, (void *)&bist_data,
-				 sizeof(struct hl_ras_sram_bist_data)))
-		preboot_puts("BOOT_INFO: Failed to set sram mbist result\n");
-	else
-		preboot_puts("BOOT_INFO: sram mbist result stored\n");
 }
 
 static bool ppboot_try_sram_mbist_data_img(struct boot_desc *desc)
@@ -212,6 +203,18 @@
 	udelay(1000);
 }
 
+static inline void gaudi3_write_sram_bist_result(void)
+{
+	/* Store Ppboot MBIST result in boot info */
+
+	if (ppboot_boot_info_set(HL_BOOT_INFO_SRAM_BIST_RESULT,
+				(void *)&bist_data,
+				sizeof(struct hl_ras_sram_bist_data)))
+		preboot_puts("BOOT_INFO: Failed to set sram mbist result\n");
+	else
+		preboot_puts("BOOT_INFO: sram mbist result stored\n");
+}
+
 void hl_sram_self_repair(void)
 {
 	struct gaudi3_cold_rst_data cold_rst_data;
@@ -261,6 +264,7 @@
 	gaudi3_sram_self_repair(sram_loadaddr);
 	gaudi3_die0_cs_reset();
 	gaudi3_sram_ecc_init();
+	gaudi3_write_sram_bist_result();
 	preboot_puts("bist done for 48MB SRAM/CACHE\n");
 }
 
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 2f9a0fc..8317717 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -116,7 +116,7 @@
  * to be fetched to the 'boot_info' structure by calling boot_info_set_no_merge() for all existing
  * descs.
  */
-static int ppboot_boot_info_set(u32 type, void *data, u32 size)
+int ppboot_boot_info_set(u32 type, void *data, u32 size)
 {
 	int ret = boot_info_sync_data();
 
diff --git a/arch/gaudi3/arc/hl.h b/arch/gaudi3/arc/hl.h
index c63dc72..5fbcb1c 100644
--- a/arch/gaudi3/arc/hl.h
+++ b/arch/gaudi3/arc/hl.h
@@ -70,5 +70,6 @@
 bool is_spi_gw_enabled(void);
 u32 gaudi3_count_set_bit(u32 n);
 void gaudi3_clean_hot_reset_intr(void);
+int ppboot_boot_info_set(u32 type, void *data, u32 size);
 
 #endif /* ARCH_GAUDI3_ARC_HL_H_ */
